File: services/query-service/src/allsource/domain/entities/projection.clj
(ns allsource.domain.entities.projection
  "Domain entities for projections.
   A projection transforms event streams into queryable read models."
  (:import [java.time Instant]))

;; ============================================================================
;; Projection Definition Entity
;; ============================================================================

(defrecord ProjectionDefinition [name version initial-state project-fn metadata])

(defrecord ProjectionSnapshot [projection-name entity-id state version timestamp])

;; ============================================================================
;; Validation Functions (Pure Domain Logic)
;; ============================================================================

(defn valid-version?
  "Check if version is valid (positive integer)."
  [version]
  (and (integer? version)
       (pos? version)))

(defn valid-name?
  "Check if projection name is valid (must be keyword)."
  [name]
  (keyword? name))

(defn valid-projection?
  "Validate a complete projection definition."
  [projection]
  (and (instance? ProjectionDefinition projection)
       (valid-name? (:name projection))
       (valid-version? (:version projection))
       (some? (:initial-state projection))
       (fn? (:project-fn projection))))

;; ============================================================================
;; Constructor Functions
;; ============================================================================

(defn make-projection
  "Create a new projection definition.

   Parameters:
   - name: Keyword identifier for the projection
   - version: Integer version number (for migration tracking)
   - initial-state: Initial state value (any data structure)
   - project-fn: Function (state, event) => new-state
   - metadata: Optional metadata map

   Example:
   (make-projection
     :name :user-statistics
     :version 1
     :initial-state {:count 0 :total-spent 0.0}
     :project-fn (fn [state event]
                   (case (:event-type event)
                     \"order.placed\" (-> state
                                          (update :count inc)
                                          (update :total-spent + (get-in event [:payload :amount])))
                     state)))"
  [& {:keys [name version initial-state project-fn metadata]
      :or {metadata {}}}]
  (map->ProjectionDefinition
    {:name name
     :version version
     :initial-state initial-state
     :project-fn project-fn
     :metadata metadata}))

(defn make-snapshot
  "Create a projection state snapshot.

   Parameters:
   - projection-name: Name of the projection
   - entity-id: ID of the entity this state belongs to
   - state: Current state value
   - version: Projection version that created this state
   - timestamp: When this snapshot was taken"
  [& {:keys [projection-name entity-id state version timestamp]}]
  (map->ProjectionSnapshot
    {:projection-name projection-name
     :entity-id entity-id
     :state state
     :version version
     :timestamp timestamp}))

;; ============================================================================
;; Projection Application (Pure Functions)
;; ============================================================================

(defn apply-event
  "Apply a single event to projection state.

   Parameters:
   - projection: ProjectionDefinition
   - state: Current state
   - event: Event to apply

   Returns: New state after applying event"
  [projection state event]
  (let [project-fn (:project-fn projection)]
    (project-fn state event)))

(defn apply-events
  "Apply multiple events to projection state.

   Parameters:
   - projection: ProjectionDefinition
   - state: Initial state
   - events: Sequence of events to apply

   Returns: Final state after applying all events"
  [projection state events]
  (reduce (partial apply-event projection) state events))

(defn initialize-state
  "Get initial state for a projection."
  [projection]
  (:initial-state projection))

;; ============================================================================
;; State Migration
;; ============================================================================

(defn migrate-state
  "Migrate projection state using a migration function.

   Parameters:
   - old-state: State from previous version
   - migration-fn: Function to transform old state to new state

   Returns: Migrated state"
  [old-state migration-fn]
  (migration-fn old-state))

;; ============================================================================
;; Accessors
;; ============================================================================

(defn get-name
  "Get projection name."
  [projection]
  (:name projection))

(defn get-version
  "Get projection version."
  [projection]
  (:version projection))

(defn get-initial-state
  "Get projection initial state."
  [projection]
  (:initial-state projection))

(defn get-metadata
  "Get projection metadata."
  [projection]
  (:metadata projection))

;; ============================================================================
;; State Snapshot Management
;; ============================================================================

(defn snapshot-state
  "Create a snapshot of current projection state.

   Parameters:
   - projection-name: Name of the projection
   - entity-id: Entity ID
   - state: Current state
   - version: Projection version

   Returns: ProjectionSnapshot"
  [projection-name entity-id state version]
  (make-snapshot
    :projection-name projection-name
    :entity-id entity-id
    :state state
    :version version
    :timestamp (Instant/now)))

(defn restore-from-snapshot
  "Restore state from a snapshot."
  [snapshot]
  (:state snapshot))

;; ============================================================================
;; Example Projections (for documentation)
;; ============================================================================

(comment
  ;; Example: User Statistics Projection
  (def user-stats-projection
    (make-projection
      :name :user-statistics
      :version 1
      :initial-state {:user-count 0
                      :total-orders 0
                      :total-revenue 0.0}
      :project-fn (fn [state event]
                    (case (:event-type event)
                      "user.created"
                      (update state :user-count inc)

                      "order.placed"
                      (-> state
                          (update :total-orders inc)
                          (update :total-revenue + (get-in event [:payload :amount])))

                      state))))

  ;; Apply events to projection
  (def events [{:event-type "user.created" :entity-id "user-1"}
               {:event-type "order.placed" :entity-id "user-1" :payload {:amount 100.0}}
               {:event-type "order.placed" :entity-id "user-1" :payload {:amount 50.0}}])

  (apply-events user-stats-projection
                (initialize-state user-stats-projection)
                events)
  ;; => {:user-count 1 :total-orders 2 :total-revenue 150.0}

  ;; Create snapshot
  (snapshot-state :user-statistics
                  "user-1"
                  {:user-count 1 :total-orders 2 :total-revenue 150.0}
                  1)
  )

File: apps/query-service/lib/query_service_ex/domain/entities/projection.ex
defmodule QueryServiceEx.Domain.Entities.Projection do
  @moduledoc """
  Domain entity for projections.

  A projection transforms event streams into queryable read models.
  This module provides pure domain logic for projection definitions and state management.
  """

  defmodule Definition do
    @moduledoc "Projection definition entity"
    @type project_fn :: (any(), map() -> any())

    @type t :: %__MODULE__{
            name: atom(),
            version: pos_integer(),
            initial_state: any(),
            project_fn: project_fn(),
            metadata: map()
          }

    @enforce_keys [:name, :version, :initial_state, :project_fn]
    defstruct [:name, :version, :initial_state, :project_fn, metadata: %{}]

    @doc """
    Create a new projection definition.

    ## Parameters
      * `name` - Atom identifier for the projection
      * `version` - Integer version number (for migration tracking)
      * `initial_state` - Initial state value (any data structure)
      * `project_fn` - Function `(state, event) -> new_state`
      * `metadata` - Optional metadata map

    ## Examples

        iex> Definition.new(
        ...>   name: :user_statistics,
        ...>   version: 1,
        ...>   initial_state: %{count: 0, total_spent: 0.0},
        ...>   project_fn: fn state, event ->
        ...>     case event.event_type do
        ...>       "order.placed" ->
        ...>         state
        ...>         |> Map.update!(:count, &(&1 + 1))
        ...>         |> Map.update!(:total_spent, &(&1 + event.payload.amount))
        ...>       _ -> state
        ...>     end
        ...>   end
        ...> )
    """
    def new(opts) do
      struct!(__MODULE__, opts)
    end
  end

  defmodule Snapshot do
    @moduledoc "Projection state snapshot entity"
    @type t :: %__MODULE__{
            projection_name: atom(),
            entity_id: String.t(),
            state: any(),
            version: pos_integer(),
            timestamp: DateTime.t()
          }

    @enforce_keys [:projection_name, :entity_id, :state, :version, :timestamp]
    defstruct [:projection_name, :entity_id, :state, :version, :timestamp]

    @doc "Create a new projection snapshot"
    def new(opts) do
      struct!(__MODULE__, opts)
    end
  end

  ## Validation Functions

  @doc "Check if version is valid (positive integer)"
  def valid_version?(version) when is_integer(version) and version > 0, do: true
  def valid_version?(_), do: false

  @doc "Check if projection name is valid (must be atom)"
  def valid_name?(name) when is_atom(name), do: true
  def valid_name?(_), do: false

  @doc "Validate a complete projection definition"
  def valid_projection?(%Definition{name: name, version: version, initial_state: initial, project_fn: project_fn})
      when is_function(project_fn, 2) do
    valid_name?(name) and valid_version?(version) and not is_nil(initial)
  end

  def valid_projection?(_), do: false

  ## Projection Application (Pure Functions)

  @doc """
  Apply a single event to projection state.

  ## Parameters
    * `projection` - ProjectionDefinition
    * `state` - Current state
    * `event` - Event to apply

  ## Returns
  New state after applying event

  ## Examples

      iex> projection = Definition.new(
      ...>   name: :counter,
      ...>   version: 1,
      ...>   initial_state: 0,
      ...>   project_fn: fn state, _event -> state + 1 end
      ...> )
      iex> Projection.apply_event(projection, 0, %{})
      1
  """
  def apply_event(%Definition{project_fn: project_fn}, state, event) do
    project_fn.(state, event)
  end

  @doc """
  Apply multiple events to projection state.

  ## Parameters
    * `projection` - ProjectionDefinition
    * `state` - Initial state
    * `events` - List of events to apply

  ## Returns
  Final state after applying all events

  ## Examples

      iex> projection = Definition.new(
      ...>   name: :counter,
      ...>   version: 1,
      ...>   initial_state: 0,
      ...>   project_fn: fn state, _event -> state + 1 end
      ...> )
      iex> Projection.apply_events(projection, 0, [%{}, %{}, %{}])
      3
  """
  def apply_events(projection, state, events) when is_list(events) do
    Enum.reduce(events, state, fn event, acc ->
      apply_event(projection, acc, event)
    end)
  end

  @doc "Get initial state for a projection"
  def initialize_state(%Definition{initial_state: initial_state}) do
    initial_state
  end

  ## State Migration

  @doc """
  Migrate projection state using a migration function.

  ## Parameters
    * `old_state` - State from previous version
    * `migration_fn` - Function to transform old state to new state

  ## Returns
  Migrated state
  """
  def migrate_state(old_state, migration_fn) when is_function(migration_fn, 1) do
    migration_fn.(old_state)
  end

  ## Accessors

  @doc "Get projection name"
  def get_name(%Definition{name: name}), do: name

  @doc "Get projection version"
  def get_version(%Definition{version: version}), do: version

  @doc "Get projection initial state"
  def get_initial_state(%Definition{initial_state: initial_state}), do: initial_state

  @doc "Get projection metadata"
  def get_metadata(%Definition{metadata: metadata}), do: metadata

  ## State Snapshot Management

  @doc """
  Create a snapshot of current projection state.

  ## Parameters
    * `projection_name` - Name of the projection
    * `entity_id` - Entity ID
    * `state` - Current state
    * `version` - Projection version

  ## Returns
  ProjectionSnapshot
  """
  def snapshot_state(projection_name, entity_id, state, version) do
    Snapshot.new(
      projection_name: projection_name,
      entity_id: entity_id,
      state: state,
      version: version,
      timestamp: DateTime.utc_now()
    )
  end

  @doc "Restore state from a snapshot"
  def restore_from_snapshot(%Snapshot{state: state}), do: state
end
